proposals/0000-decompose-newtypes.rst

```diff
@@ -0,0 +1,182 @@
+For a newtype ``T``, given ``[W] (T t1) ~R (T t2)``
+can we ever decompose to ``[W] t1 ~R t2``, without losing completeness?
+
+The relevant paper is `Safe zero-cost coercions for Haskell (journal version) <https://simon.peytonjones.org/safe-coercions-2016/>`_.
+Here's the CO-NTH rule in the paper, Fig 5::
+
+  G |- g : H taus ~R H sigmas    rs is a prefix of roles(H)     H is not a newtype
+  ---------------------------------------------------------------------------------------    CO-NTH
+     G |- Nth(i) G : taus(i) ~rs(i) sigmas(i)
+
+The CO-NTH rule governs completeness of decomposition for wanted constraints: if CO-NTH holds
+for each argument i, it means that nothing is lost by decomposing.
+
+But note "H is not a newtype"; that means **we can never decompose a newtype** except
+at risk of incompleteness.
+But in the case of the newtype ``IO`` above, it is easy to see that
+if we know ``(IO t1) ~R (IO t2)`` then we certainly know ``t1 ~R t2``, and vice versa.
+**Nothing is lost by decomposing ``IO``!**  CO-NTH is too conservative.
+
+The proposal below (see Proposed Change Specification) strengthens CO-NTH to allow
+this decomposition.
+
+
+
+Proposed Change Specification
+-----------------------------
+
+There are two proposed changes.
+
+Change 1: CO-NTH Rule
+:::::::::::::::::::::
+
+First, change the CO-NTH rule as follows::
+
+  G |- g : H taus ~R H sigmas    rs is a prefix of roles(H)
+  H is not a newtype or rs(i) is representational
+  ---------------------------------------------------------------------------------------    CO-NTH
+     G |- Nth(i) G : taus(i) ~rs(i) sigmas(i)
+
+Note the extra "or rs(i) is representational".  That says that if we know ``(T ty1) ~R (T ty2)``
+then we know ``ty1 ~R ty2`` *if ``T``'s argument has representational role*. When is that?
+Here's an example::
+
+  type family F
+  newtype N a b c = MkN ([a], F c)
+
+  --                   a            b      c
+  -- type role N representational phantom nominal
+
+Here
+* ``b`` has phantom role: it is not even mentioned in the RHS
+* ``c`` has nominal rule: it appears in the RHS but only in the argument of a type family
+* ``a`` has reprsentational role: it appears in the RHS, and *not* under a type family
+
+Representational role means "if you know the RHS type, then you know ``a``"
+
+Change 2: user role signatures
+::::::::::::::::::::::::::::::
+
+In GHC today, the user can declare explicit roles::
+
+  newtype P a = MkP Int
+  type role P representational
+
+Here the argument of ``P`` is not mentioned in its RHS, so its role would be inferres as ``phantom``.  But GHC today allows you to override a phantom role
+to ``representational``.
+
+If we continue to allow this, the new CO-NTH rule would be unsound.  Clearly, knowing ``(P t1) ~R (P t2)`` does **not** imply that ``t1 ~R t2``.
+
+Conclusion: **the new CO-NTH requires that we do not allow a user to give a representational role for a phantom argument.**  Any such
+attempt would simply be rejected.
```

Comment:
> To be clear, I take it that this restriction applies only to newtypes, not role ascriptions on datatypes?

Yes indeed.  I will clarify.

>  What about role ascriptions on classes (potentially represented via newtypes if they have a single method)?

Huh -- that is a fine question.  I think classes should be treated uniformly, unaffected the implementation choice to implement them with a newtype. I'm not confident that GHC does so.  But equalities *between class constraints* are rare... like `Eq t1 ~ Eq t2`. I'm not sure they can occur at all, actually.  Maybe with impredicative types....  So, very much a corner case, and one that might deserve study outside this proposal.

> What about type role P nominal, is that allowed?

Yes you can upgrade to `nominal`.

> But it would feel strange if the definition of P could not be given the "weaker" role representational but could be given the "stronger" role nominal.

Ah, but `representational` is very strong indeed!  It is not weaker (in this sense).

Reply:
Thanks for clarifying!

> > What about role ascriptions on classes (potentially represented via newtypes if they have a single method)?
> 
> Huh -- that is a fine question. I think classes should be treated uniformly, unaffected the implementation choice to implement them with a newtype. I'm not confident that GHC does so. But equalities between class constraints are rare... like `Eq t1 ~ Eq t2`. I'm not sure they can occur at all, actually. Maybe with impredicative types.... So, very much a corner case, and one that might deserve study outside this proposal.

It's certainly a strange corner. I've only once used a `representational` role annotation on a class (https://well-typed.com/blog/2015/07/checked-exceptions/), and in that case the dictionary would be a unit datatype with a `phantom` parameter, as it happens.

Reply:
I think that in example below occurs `Eq a ~ Eq b` constraint.

```haskell
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE ImpredicativeTypes #-}
module Constr where

import Data.Type.Equality
import Data.Type.Coercion

{-
-- RHS size: {terms: 11, types: 64, coercions: 6, joins: 0/0}
promote'
  :: forall a b.
     Coercion (Eq a => a -> a -> Bool) (Eq b => b -> b -> Bool)
     -> a :~: b
[GblId, Arity=1, Unf=OtherCon []]
promote'
  = \ (@a_aNJ)
      (@b_aNK)
      (ds_dPd
         :: Coercion
              (Eq a_aNJ => a_aNJ -> a_aNJ -> Bool)
              (Eq b_aNK => b_aNK -> b_aNK -> Bool)) ->
      case ds_dPd of { Coercion $dCoercible_aNN ->
      case GHC.Types.coercible_sel
             @(*)
             @(Eq a_aNJ => a_aNJ -> a_aNJ -> Bool)
             @(Eq b_aNK => b_aNK -> b_aNK -> Bool)
             $dCoercible_aNN
      of co_aO5
      { __DEFAULT ->
      (Data.Type.Equality.$WRefl @(*) @a_aNJ)
      `cast` (((:~:) <*>_N <a_aNJ>_N (Nth:0 (Nth:3 co_aO5)))_R
              :: (a_aNJ :~: a_aNJ) ~R# (a_aNJ :~: b_aNK))
      }
      }

-}
promote' :: Coercion (Eq a => a -> a -> Bool) (Eq b => b -> b -> Bool) -> a :~: b
promote' Coercion = Refl
```

Reply:
Hmm.... my guess is that GHC gets this wrong today, in that I imagine it will decompose a newtype-class. We should check. ...checks.... Actually, GHC does get this right. `class C a where deflt :: a` does not decompose, whereas `class C2 a where deflt :: a; blah :: a -> a` does. So, nothing to see here.